### PR TITLE
ADCM-6158: Implement missing methods for HostNode

### PR DESCRIPTION
**Pull request by Starovoitov**

### CODE REVIEW

File: adcm_aio_client/core/objects/cm.py [hunk 1]
```diff
@@ -297,6 +304,14 @@ class HostsAccessor(PaginatedAccessor[Host, None]):
     class_type = Host
 
 
+class HostsNode(HostsAccessor):
+    async def add(self: Self, provider: HostProvider, name: str, cluster: Cluster | None = None) -> None:
+        data = {"hostprovider_id": provider.id, "name": name}
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:309]: don't use internal formats pls
`hostproviderId`, `clusterId`

File: adcm_aio_client/core/types.py [hunk 2]
```diff
@@ -84,3 +84,9 @@ class ADCMEntityStatus(str, Enum):
 class MappingOperation(str, Enum):
     ADD = "add"
     REMOVE = "remove"
+
+
+class MaintenanceModeStatus(str, Enum):
+    ON = "on"
+    OFF = "off"
+    CHANGING = "changing"
```
Comment (Sealwing) [inline adcm_aio_client/core/types.py:92]: move to types

File: adcm_aio_client/core/objects/_base.py [hunk 3]
```diff
@@ -84,3 +90,24 @@ def __init__(self: Self, parent: Parent, requester: Requester, data: dict[str, A
 
     def get_own_path(self: Self) -> Endpoint:
         return *self._parent.get_own_path(), self.PATH_PREFIX, self.id
+
+
+class MaintenanceMode:
+    def __init__(
+        self: Self, maintenance_mode_status: MaintenanceModeStatus, requester: Requester, path: Endpoint
+    ) -> None:
+        self.maintenance_mode_status = maintenance_mode_status
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/_base.py:99]: make protected, it should be available to be changed publicly 

you may add propery like `value`, so user can access it

File: adcm_aio_client/core/objects/_base.py [hunk 4]
```diff
@@ -84,3 +90,24 @@ def __init__(self: Self, parent: Parent, requester: Requester, data: dict[str, A
...
+        self._path = path
+
+    def __repr__(self: Self) -> MaintenanceModeStatus:
+        return self.maintenance_mode_status
+
+    def __str__(self: Self) -> MaintenanceModeStatus:
+        return self.maintenance_mode_status
+
+    async def on(self: Self) -> None:
+        await self._requester.post(*self._path, "maintenance-mode", data={"maintenanceMode": MaintenanceModeStatus.ON})
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/_base.py:110]: update internal value

note that it may become `changing` instead of `on`

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py [hunk 5]
```diff
@@ -341,6 +348,14 @@ class HostsAccessor(PaginatedAccessor[Host, None]):
     class_type = Host
 
 
+class HostsNode(HostsAccessor):
+    async def add(self: Self, provider: HostProvider, name: str, cluster: Cluster | None = None) -> None:
+        data = {"hostproviderId": provider.id, "name": name}
+        if cluster:
+            data["clusterId"] = cluster.id
+        await self._requester.post(*self._path, data=data)
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:356]: always use `get_own_path`, not implementation details, pls

File: adcm_aio_client/core/objects/cm.py [hunk 6]
```diff
@@ -333,6 +334,12 @@ async def cluster(self: Self) -> Cluster | None:
     async def hostprovider(self: Self) -> HostProvider:
         return await HostProvider.with_id(requester=self._requester, object_id=self._data["hostprovider"]["id"])
 
+    @async_cached_property
+    async def maintenance_mode(self: Self) -> MaintenanceMode:
+        response = await self._requester.get(*self.get_own_path())
+        self._data = response.as_dict()
```
Comment (Sealwing) [inline adcm_aio_client/core/objects/cm.py:340]: don't reread on cached property request

there's a `refresh` mechanism for all of that

MM should be the one retrieved on host retrieval from API

User will refresh object when they want "current" value

---

It's also a good idea to add `WithMaintenanceMode` mixin

**Review by Sealwing — CHANGES_REQUESTED** [review]

File: adcm_aio_client/core/objects/cm.py (hunk 5 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py:351]: It's a node. It doesn't have get_own_path

File: adcm_aio_client/core/objects/cm.py (hunk 6 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: added

File: adcm_aio_client/core/objects/_base.py (hunk 3 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/_base.py]: done

File: adcm_aio_client/core/objects/_base.py (hunk 4 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/_base.py:114]: done

Comment (Sealwing) [reply adcm_aio_client/core/objects/_base.py:114]: nope, see about `changing`

File: adcm_aio_client/core/objects/cm.py (hunk 6 above)
Comment (Sealwing) [reply adcm_aio_client/core/objects/cm.py]: no, it's not what I've meant

User should call refresh when they feel adequate, not we calling it or any similar code in here

pls use existing data for that, no requests

Comment (Starovoitov) [reply adcm_aio_client/core/objects/cm.py]: changed

File: adcm_aio_client/core/objects/_base.py (hunk 4 above)
Comment (Starovoitov) [reply adcm_aio_client/core/objects/_base.py:114]: changed

File: adcm_aio_client/core/objects/cm.py [hunk 7]
```diff
@@ -340,6 +341,14 @@ class HostsAccessor(PaginatedAccessor[Host, None]):
     class_type = Host
 
 
+class HostsNode(HostsAccessor):
+    async def add(self: Self, provider: HostProvider, name: str, cluster: Cluster | None = None) -> None:
```
Comment (a-alferov) [inline adcm_aio_client/core/objects/cm.py:345]: * I think, this method should be `create()`
* Add `HostsNode` to `Client`
* add `description` field to method

**Review by Sealwing — APPROVED** [review]

**Review by a-alferov — APPROVED** [review]

**Pull request merged**